Add table of branch comparisons (#108)
ignitionrobotics/docs: tools/branch_comparisons.md

# Branch comparisons

These links take to comparisons between branches.

If a cell in the table is `-`, this means that the two Ignition releases use
the same version of this Ignition library. So, for example, `ign-cmake` has a
`-` for B ➡️  C because Blueprint and Citadel use `ign-cmake2`.

> When it says "There isn’t anything to compare," this means no pull requests

Reviewer: ```suggestion
When the comparison link says "There isn’t anything to compare," this means no pull requests
```

"it" is a little vague here, so I think it would be good to clarify that we are referring to the branch comparison, and not the table itself.

Author: a667d98

need to be forward-ported.

Otherwise, those branches could use a forward-port pull request.
See the instructions on the
[contributing guide](https://ignitionrobotics.org/docs/all/contributing#process).

* **B**: Blueprint
* **C**: Citadel
* **D**: Dome
* **main**: main branch

> SDFormat is a special case where the release branches should be compared
  against `merge_N_M_reference` branches instead of other release branches.

>   Be sure to update those branches before using for comparison.

Reviewer: Do we still need this if SDFormat is following the Ignition porting strategy now?

Author: a667d98

# Branch comparisons

These links take to comparisons between branches.

If a cell in the table is `-`, this means that the two Ignition releases use
the same version of this Ignition library. So, for example, `ign-cmake` has a
`-` for B ➡️  C because Blueprint and Citadel use `ign-cmake2`.

When it says "There isn’t anything to compare," this means no pull requests
need to be forward-ported.

> Otherwise, those branches could use a forward-port pull request, see
> instructions on the

Reviewer: ```suggestion
Otherwise, those branches could use a forward-port pull request.
See the instructions on the
```

[contributing guide](https://ignitionrobotics.org/docs/all/contributing#process).

* **B**: Blueprint
* **C**: Citadel
* **D**: Dome
* **main**: main branch

> > SDFormat is a special case where the release branches should be compared
>   against `merge_N_M_reference` branches instead of other release branches.
>   Be sure to update those branches before using for comparison.

Reviewer: ```suggestion
> SDFormat is a special case that doesn't follow the forward-port process listed in the contributing guide.
  For SDFormat, the release branches should be compared against `merge_N_M_reference` branches (where `M = N + 1`) instead of other release branches.
  Be sure to update the `merge_N_M_reference` branches before using them for comparison.
  The `merge_N_M_reference` branches need to be updated if `compare/merge_N_M_reference...sdfN` isn't empty.
  If this comparison isn't empty, update `merge_N_M_reference` by merging `sdfN` into `merge_N_M_reference`.
  After updating `merge_N_M_reference`, see what's in `merge_N_M_reference` that hasn't been merged forward to `sdfM` yet by looking at the **diff** of `compare/sdfM...merge_N_M_reference` (the list of commits in this comparison isn't helpful), and cherry-pick accordingly.
```

I tried to clarify this a little more based on this [comment](https://github.com/ignitionrobotics/docs/pull/108#discussion_r509484526), but I'm not 100% sure I understand how the cherry-picking part works - do we cherry-pick commits into a separate branch that is then used to create a forward-port PR?

Author: SDFormat is using the same pattern as the rest now, so this isn't needed.

# Branch comparisons

These links take to comparisons between branches.

> When it says "There isn’t anything to compare," this means no pull requests
> need to be forward-ported.

Reviewer: ```suggestion
If a cell in the table is `-`, this means that the two Ignition releases use the same version of this Ignition library.
So, for example, `ign-cmake` has a `-` for B ➡️  C because Blueprint and Citadel use `ign-cmake2`.

When it says "There isn’t anything to compare," this means no pull requests
need to be forward-ported.
```

Author: 824176b170a8353a14f69bde100af7982fd8aa53

Otherwise, those branches could use a forward-port pull request, see
instructions on the
[contributing guide](https://ignitionrobotics.org/docs/all/contributing#process).

* **B**: Blueprint
* **C**: Citadel
* **D**: Dome
* **main**: main branch

> Library | B ➡️  C | C ➡️  D | ➡️  main

Reviewer: ```suggestion
Library | B ➡️  C | C ➡️  D | D ➡️  main
```
When I looked at this table originally, it wasn't super clear to me what version of each repository was being compared to `main` until I clicked on the `➡️  main` link. It looks like the `➡️  main` column is comparing the most recent ignition release (in this case, `Dome`) to `main`, which makes sense - perhaps we should explicitly state this in the table so it's clear to all readers?

Author: Good point, I don't remember exactly what my reasoning was to left that out but I think that makes sense. Changed in 824176b170a8353a14f69bde100af7982fd8aa53

[ign-gazebo][ign-gazebo] | [2 ➡️  3][ign-gazebo-2-3] | [3 ➡️  4][ign-gazebo-3-4] | [➡️  main][ign-gazebo-main]
[ign-gui][ign-gui] | [2 ➡️  3][ign-gui-2-3] | [3 ➡️  4][ign-gui-3-4] | [➡️  main][ign-gui-main]
[ign-launch][ign-launch] | [1 ➡️  2][ign-launch-1-2] | [2 ➡️  3][ign-launch-2-3] | [➡️  main][ign-launch-main]
[ign-math][ign-math] | - | - | [➡️  main][ign-math-main]
[ign-msgs][ign-msgs] | [4 ➡️  5][ign-msgs-4-5] | [5 ➡️  6][ign-msgs-5-6] | [➡️  main][ign-msgs-main]
[ign-physics][ign-physics] | [1 ➡️  2][ign-physics-1-2] | [2 ➡️  3][ign-physics-2-3] | [➡️  main][ign-physics-main]
[ign-plugin][ign-plugin] | - | - | [➡️  main][ign-plugin-main]
[ign-rendering][ign-rendering] | [2 ➡️  3][ign-rendering-2-3] | [3 ➡️  4][ign-rendering-3-4] | [➡️  main][ign-rendering-main]
[ign-sensors][ign-sensors] | [2 ➡️  3][ign-sensors-2-3] | [3 ➡️  4][ign-sensors-3-4] | [➡️  main][ign-sensors-main]
[ign-tools][ign-tools] | - | - | [➡️  main][ign-tools-main]
[ign-transport][ign-transport] | [7 ➡️  8][ign-transport-7-8] | [8 ➡️  9][ign-transport-8-9] | [➡️  main][ign-transport-main]

> 

Reviewer: Do we want to make any mention of `sdformat`, or does that not apply here since that's not a part of the Ignition org?

Author: The reason I left SDFormat out is because that repository hasn't been using the same merge strategy. Changes are being cherry-picked then rebase-merged on all directions (forwards and backwards). That makes the GitHub comparison a bit crazy. For example see the diff from `sdf8` to `sdf9`:

https://github.com/osrf/sdformat/compare/sdf9...sdf8

That shows for example as if `Material::FilePath` wasn't on `sdf9` yet. Which is not true, because it's there :woman_shrugging: :

https://github.com/osrf/sdformat/blob/sdf9/src/Material.cc#L355-L359

@scpeters , you've been cherry-picking those changes. How do you keep track of what needs porting? It would be good to document that here too so all maintainers can help out. Ideally, we'd use the same strategy for all libraries, but I understand SDFormat has been following a slightly different process.

Reviewer 2: Yes, I've been cherry-picking instead of merging. I use the following branches to keep track of outdated changes:

* https://github.com/osrf/sdformat/tree/merge_8_9_reference
* https://github.com/osrf/sdformat/tree/merge_9_10_reference

I regularly merge `sdf{X}` and `sdf{X+1}` into these branches and resolve conflicts. To see what needs to be merged forward from `sdf9` to `sdf10` for example:

1. Make sure that you've merged forward from `sdf9` to `merge_9_10_reference` by ensuring that [compare/merge_9_10_reference...sdf9](https://github.com/osrf/sdformat/compare/merge_9_10_reference...sdf9) is empty; if not, then merge `sdf9` into that branch.

2. Check to see what is in `merge_9_10_reference ` that hasn't yet been merged forward to `sdf10`, by looking at the diff of [compare/sdf10...merge_9_10_reference](https://github.com/osrf/sdformat/compare/sdf10...merge_9_10_reference#files_bucket) (the list of commits isn't helpful).

At the moment that I type this, both of those comparisons are not empty, meaning that there are changes from `sdf9` that need to be merged forward to `merge_9_10_reference`, and changes in `merge_9_10_reference` that aren't yet in `sdf10`.

I'm not sure if other people like this workflow or not, but it's kept the sdformat release branches pretty clean so far.

Author: I took a stab at documenting SDF's case in 2c1b8e88697fa8376fcafeb1accbe854f6e8b7a0. Feel free to make edits directly to this branch to improve it.

> I'm not sure if other people like this workflow or not, but it's kept the sdformat release branches pretty clean so far.

I must say I'm not convinced that this workflow is scalable and maintainable in the long run. There are lots of moving pieces to be kept in sync, and as you mentioned, the diff commit list isn't helpful. Sure, it keeps a linear history, but at a high maintenance cost. Is that worth it? It may be suitable for a low-traffic repository that has someone keeping a close eye at every change, but I think it will become more difficult as hundreds of changes get intertwined across different branches. I think it also makes it more difficult to share the responsibility across more people in the team.

That said, I'm not opposed to keeping the status quo a bit longer to see how it goes. I think it's good to do some experimentation. The process we're doing for the Ignition libraries is far from perfect. There are extra merge commits sprinkled all over the history, and it's harder to track what was merged when. I'm all ears for improvements. But in my opinion, so far, that's the option that scales better across several libraries with several versions and several maintainers.

Reviewer 2: after talking about it with @azeey, I'm giving up the cherry-pick pattern with sdformat and have opened a PR to start merging forward from sdf8 -> sdf9 and will continue with the other branches: https://github.com/osrf/sdformat/pull/445

so sdformat can be treated like the others with consistency

Author: Thanks for iterating, @scpeters . Updated in 81dd4f674d0fc7f9a78667760e3becf281a45ce4.
